pkg/virt-operator/install-strategy/create.go

```diff
@@ -0,0 +1,995 @@
+
+		_, err = apps.Deployments(kv.Namespace).Patch(deployment.Name, types.JSONPatchType, generatePatchBytes(ops))
+		if err != nil {
+			return fmt.Errorf("unable to patch deployment %+v: %v", deployment, err)
+		}
+		log.Log.V(2).Infof("deployment %v updated", deployment.GetName())
+
+	} else {
+		log.Log.V(4).Infof("deployment %v is up-to-date", deployment.GetName())
+	}
+
+	return nil
+}
+
+func shouldTakeUpdatePath(targetVersion, currentVersion string) bool {
+
+	// if no current version, then this can't be an update
+	if currentVersion == "" {
+		return false
+	}
+
+	// semver doesn't like the 'v' prefix
+	targetVersion = strings.TrimPrefix(targetVersion, "v")
+	currentVersion = strings.TrimPrefix(currentVersion, "v")
+
+	// our default position is that this is an update.
+	// So if the target and current version do not
+	// adhere to the semver spec, we assume by default the
+	// update path is the correct path.
+	shouldTakeUpdatePath := true
+	target, err := semver.Make(targetVersion)
+	if err == nil {
+		current, err := semver.Make(currentVersion)
+		if err == nil {
+			if target.Compare(current) <= 0 {
+				shouldTakeUpdatePath = false
+			}
+		}
+	}
+
+	return shouldTakeUpdatePath
+}
+
+func SyncAll(kv *v1.KubeVirt,
+	prevStrategy *InstallStrategy,
+	targetStrategy *InstallStrategy,
+	stores util.Stores,
+	clientset kubecli.KubevirtClient,
+	expectations *util.Expectations) (bool, error) {
+
+	kvkey, err := controller.KeyFunc(kv)
+	if err != nil {
+		return false, err
+	}
+
+	gracePeriod := int64(0)
+	deleteOptions := &metav1.DeleteOptions{
+		GracePeriodSeconds: &gracePeriod,
+	}
+
+	ext := clientset.ExtensionsClient()
+	core := clientset.CoreV1()
+	rbac := clientset.RbacV1()
+	scc := clientset.SecClient()
+
+	imageTag := kv.Status.TargetKubeVirtVersion
+	imageRegistry := kv.Status.TargetKubeVirtRegistry
+
+	takeUpdatePath := shouldTakeUpdatePath(kv.Status.TargetKubeVirtVersion, kv.Status.ObservedKubeVirtVersion)
+
+	// -------- CREATE AND ROLE OUT UPDATED OBJECTS --------
+
+	// create/update CRDs
+	for _, crd := range targetStrategy.crds {
+		var cachedCrd *extv1beta1.CustomResourceDefinition
+
+		crd := crd.DeepCopy()
+		obj, exists, _ := stores.CrdCache.Get(crd)
+		if exists {
+			cachedCrd = obj.(*extv1beta1.CustomResourceDefinition)
+		}
+
+		injectOperatorLabelAndAnnotations(&crd.ObjectMeta, imageTag, imageRegistry)
+		if !exists {
+			// Create non existent
+			expectations.Crd.RaiseExpectations(kvkey, 1, 0)
+			_, err := ext.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd)
+			if err != nil {
+				expectations.Crd.LowerExpectations(kvkey, 1, 0)
+				return false, fmt.Errorf("unable to create crd %+v: %v", crd, err)
+			}
+			log.Log.V(2).Infof("crd %v created", crd.GetName())
+
+		} else if !objectMatchesVersion(&cachedCrd.ObjectMeta, imageTag, imageRegistry) {
+			// Patch if old version
+			var ops []string
+
+			if crd.Spec.Version != cachedCrd.Spec.Version {
```

Comment:
I think you have to check all CRDs upfront, to avoid creating partially updated CRDs.

Reply:
this is a problem for all our resources.  Right now, we can restore to the previous version by simply reverting the imageTag in the KubeVirt spec. It will always converge back even if we were in a partial state.